snuba/subscriptions/scheduler_consumer.py

```diff
@@ -0,0 +1,279 @@
+                )
+            except InvalidRangeError:
+                logger.warning(
+                    "Could not construct valid time interval between %r and %r!",
+                    previous_message,
+                    MessageDetails(commit.offset, commit.orig_message_ts),
+                    exc_info=True,
+                )
+                return None
+            else:
+                result = Message(
+                    message.partition,
+                    message.offset,
+                    Tick(
+                        commit.partition.index,
+                        Interval(previous_message.offset, commit.offset),
+                        time_interval,
+                    ).time_shift(self.__time_shift),
+                    message.timestamp,
+                )
+        else:
+            result = None
+
+        self.__previous_messages[commit.partition] = MessageDetails(
+            commit.offset, commit.orig_message_ts
+        )
+
+        return result
+
+    def pause(self, partitions: Sequence[Partition]) -> None:
+        self.__consumer.pause(partitions)
+
+    def resume(self, partitions: Sequence[Partition]) -> None:
+        self.__consumer.resume(partitions)
+
+    def paused(self) -> Sequence[Partition]:
+        return self.__consumer.paused()
+
+    def tell(self) -> Mapping[Partition, int]:
+        return self.__consumer.tell()
+
+    def seek(self, offsets: Mapping[Partition, int]) -> None:
+        self.__consumer.seek(offsets)
+
+    def stage_positions(self, positions: Mapping[Partition, Position]) -> None:
+        return self.__consumer.stage_positions(positions)
+
+    def commit_positions(self) -> Mapping[Partition, Position]:
+        return self.__consumer.commit_positions()
+
+    def close(self, timeout: Optional[float] = None) -> None:
+        return self.__consumer.close(timeout)
+
+    @property
+    def closed(self) -> bool:
+        return self.__consumer.closed
+
+
+class SchedulerBuilder:
+    def __init__(
+        self,
+        entity_name: str,
+        partitions: Optional[int],
+        consumer_group: str,
+        auto_offset_reset: str,
+        delay_seconds: Optional[int],
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__entity = get_entity(EntityKey(entity_name))
+
+        storage = self.__entity.get_writable_storage()
+
+        assert (
+            storage is not None
+        ), f"Entity {entity_name} does not have a writable storage by default."
+
+        stream_loader = storage.get_table_writer().get_stream_loader()
+
+        commit_log_topic_spec = stream_loader.get_commit_log_topic_spec()
+        assert commit_log_topic_spec is not None
+        self.__commit_log_topic_spec = commit_log_topic_spec
+
+        self.__partitions = (
+            partitions
+            if partitions is not None
+            else stream_loader.get_default_topic_spec().partitions_number
+        )
```

Comment:
This part is just the same as how it works today https://github.com/getsentry/snuba/blob/master/snuba/cli/subscriptions.py#L209-L213. I can remove the option to override via CLI if you like.

Reply:
Yes, today it works like that, but I would like us to fix this and stop depending on a hardcoded number of partitions for this scheduler. 
Does not necessarily need to be fixed in this pr but let's ensure we improve this.

Reply:
We aren't depending much on the hardcoded partitions anymore. It plays a much less important role now than it did before. Now it is only used know the total number of partitions that we are going to wait for to determine the slowest one. If the scheduler ends up being single partitioned then we can find a way to estimate this automatically later, otherwise partition information of some sort will probably need to be provided. The initial goal of this scheduler implementation is to actually collect information and do load testing in order to decide on the partitioning strategy.